### ISSUE TITLE
Improvement: ability to "type hint" inputs to python node or code block

### ISSUE THREAD

**Issue by AndrewHeumann** [DynamoDS/Dynamo#6984]
If this issue is with Dynamo for Revit, please post your issue on the [Dynamo for Revit Issues page](https://github.com/DynamoDS/DynamoRevit/issues).

If this issue is **not** a bug report or improvement request, please check the [Dynamo forum](https://forum.dynamobim.com/), and start a thread there to discuss your issue.

## Dynamo version

1.1.0

## Operating system

win 7

## Request

When I am writing scripts in python or design script, I try to write them as generally as possible - so that they work on items, or lists, or lists of lists. I do this by writing them to operate on a single item and then making them a custom node, where I can specify a type hint of :var or :<Type> so that dynamo will handle iterating over the list if I pass it one. (Contrast this to many Clockwork nodes, which wrap each python script in a "Turn into list" / "Return List or single value" workflow which frequently fails and cannot be properly list.mapped over a sequence.)

This requires a somewhat tedious sequence in practice - where I'm trying to write my script this way but have a list to work with, so I have to:
1. grab the first item on my list
2. write the script to function properly on that item
3. wrap the script in a custom node
4. run it again over the entire list
And then, if there's a bug or failure, custom node reporting of internal errors is spotty at best - to debug I usually have to go into the custom node, copy the script, paste it back in the document, find the item causing the error, test it on this item, and when solved update the custom node.

ALLL THIS is to say - 
if python scripts (or code block nodes, by extension) had an interface that exposed a type hint, much of this hassle could be avoided.  It would also serve to elevate what I think is a critical best practice for reusability from a "sneaky trick" into a first-class feature. 

Grasshopper scripting nodes all have this ability - and it is _very_ useful. 
![image](https://cloud.githubusercontent.com/assets/2157902/17378296/5a0a81fc-598b-11e6-82ef-8d871da1aaf4.png)

**Comment by ksobon** [comment]
I agree with @AndrewHeumann that this would be helpful. I personally implement list handling in my Python code directly so I don't have to deal with this "workflow" at all, but then again its not how it should be done from a purist stand point of view. 

ps. @AndrewHeumann I think that Clockwork ( @andydandy74 ) was made that way because its VERY OLD. His implementation of "Turn Into List" is a leftover from early days of Dynamo (6.3, 7.3 etc) that didn't allow to specify var: input type on Custom Nodes at all. That's why I have always been handling it in Python code, and Andy with his trick.
